### PR TITLE
feat(finetune): add response-only chat and ShareGPT support

### LINKED ISSUE #823: [REQ] add response-only chat and ShareGPT support
Parent: #214

Depends on #598 and #822.

## Summary

Add explicit response-only SFT for `type: messages` and add `type: sharegpt` as a deterministic compatibility adapter.

```yaml
datasets:
  - source: acme/conversations
    type: messages

config:
  unsloth:
    loss: response
```

ShareGPT input:

```json
{
  "conversations": [
    {"from": "human", "value": "What is 2+2?"},
    {"from": "gpt", "value": "4"}
  ]
}
```

## Motivation

Native `assistant_only_loss` is not a reliable default for the frozen stack because it requires tokenizer templates that emit assistant masks, generally through `{% generation %}` regions. Tested Qwen and Llama templates did not reliably provide those masks.

Unsloth's `train_on_responses_only` is the practical chat response-only path, provided marker detection and resulting labels are validated. Unsloth's generic ShareGPT standardizer is heuristic and can fail on valid one-row or metadata-rich datasets, so AIKit should own an explicit adapter.

## Proposed changes

- Add an SFT loss option:
  - `all` — existing full-sequence chat behavior.
  - `response` — supported initially for `messages` and `sharegpt`.
- Keep `all` as the default for chat datasets.
- Reuse canonical message validation and chat-template rendering from #822.
- For `loss: response`:
  - Apply Unsloth response-only masking after rendering.
  - Validate that response markers can be derived.
  - Fail if any prepared training set has no supervised response tokens.
  - Never silently fall back to full-sequence loss.
- Add `type: sharegpt` and normalize deterministically:

```text
system             -> system
human, user        -> user
gpt, assistant     -> assistant
```

- Require string `from` and `value` fields.
- Reject unknown roles and non-string content.
- Do not use heuristic key inference as the public contract.
- Avoid logging full dataset URLs or sensitive query strings.

## Acceptance criteria

- `messages` with `loss: response` supervises assistant response tokens only.
- Omitted loss or `loss: all` retains full-sequence behavior.
- A valid one-row ShareGPT dataset normalizes successfully.
- ShareGPT records use the same rendering and loss pipeline as messages.
- Missing chat templates, unknown roles, malformed content, or conversations without assistant responses fail clearly.
- A fully masked dataset fails before `trainer.train()`.
- Existing Alpaca, prompt-completion, text, and full-sequence messages behavior remains unchanged.
- Documentation explains full-sequence versus response-only chat training.
- A one-step response-only GPU smoke run exports a valid GGUF.

## Tests

- Deterministic role-mapping tests, including a one-row ShareGPT regression.
- Tests for unknown roles, missing fields, empty conversations, and non-string content.
- Tests verifying `add_generation_prompt=False` and no Alpaca formatting.
- Missing/unusable chat-template tests.
- Label-level test proving user/prompt tokens are masked and response tokens remain supervised.
- Test proving a completely masked dataset fails.
- Exact-lock test with a tokenizer fixture containing known user/assistant markers.
- GPU fine-tune/export smoke test.

## Out of scope

- Native `assistant_only_loss`.
- User-provided Jinja templates or marker strings.
- Tool calls, `tool` messages, or multimodal content.
- Streaming datasets.
- Per-dataset loss settings.
- DPO, ORPO, KTO, GRPO, vision, or audio.

### PR DESCRIPTION
**Pull request by sozercan**
Stacked on #831. This PR targets `issue-822` and should be merged after #831.

## Summary

- add `config.unsloth.loss: all|response`, defaulting to existing full-sequence chat behavior
- add deterministic `sharegpt` normalization for `system`, `human|user`, and `gpt|assistant` roles
- apply the pinned Unsloth response-only masking path for messages and ShareGPT datasets
- reject response-only packing so masks cannot cross conversation boundaries
- validate marker derivation, marker collisions, prepared labels, and fully masked datasets before training
- preserve existing Alpaca, prompt-completion, text, and full-sequence messages behavior
- add messages-response and ShareGPT-response GPU smoke modes, fixtures, and documentation

## Tests

- `make test`
- `make test-python` (110 tests)
- `go test ./pkg/aikit/config ./pkg/aikit2llb/finetune ./pkg/build -count=1`
- exact-lock check with uv 0.12.1
- changed-lines `golangci-lint` (0 issues)
- workflow YAML and JSONL fixture validation
- `go mod tidy` clean
- independent pinned-Unsloth API review

The GPU fine-tune job will be run on the completed top-of-stack branch.

Closes #823

## GPU validation

- `sharegpt-response-smoke` passed end-to-end on the descendant `issue-825` branch: [run 31174322784](https://github.com/kaito-project/aikit/actions/runs/31174322784).

### CODE REVIEW

File: pkg/finetune/target_unsloth.py [hunk 1]
```diff
@@ -21,26 +30,99 @@
...
+    DATASET_LOADER_PARQUET: frozenset((".parquet",)),
+    DATASET_LOADER_TEXT: frozenset((".txt", ".text")),
+}
+REMOTE_DATASET_COMPRESSION_SUFFIXES = frozenset(
+    (".bz2", ".gz", ".xz", ".zip", ".zst")
+)
+REMOTE_DATASET_CHUNK_SIZE = 1024 * 1024
 DATASET_TYPE_ALPACA = "alpaca"
 DATASET_TYPE_MESSAGES = "messages"
+DATASET_TYPE_PREFERENCE = "preference"
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:80]: This expanded diff is an artifact of the final stacked merge, not an unreviewed scope expansion in the original PR. Each downstream feature remained in its own PR with independent review, unit tests, documentation, and live GPU validation (#831 through #835). During landing, the already-reviewed child branches were merged upward through their feature bases so the stack could be merged into `main`; that made the lower PR's final branch contain the downstream commits. No additional implementation change is needed for this scope-only thread.
